### PR TITLE
[Event Hubs Client] Capture Unobserved AMQP Link Close Errors

### PR DESCRIPTION
**Pull request by jsquire**
# Summary 

The AMQP transport library used by Event Hubs manages aspects of the transport in the background, notably the flow of data into the prefetch queue.  In doing
so it controls attempts to read data, and detecting connection/link failures to provide resiliency for callers.

Access to the AMQP link is managed through the `FaultTolerantAmqpObject<T>` which uses a `GetOrCreate` pattern to handle faults and provide a resilient experience for callers.  When a link was discarded due to a fault, it was not observable to clients and a new link would be created on the next `GetOrCreate` call.  In the majority of scenarios, this behavior is desirable.  However, in some scenarios - such as a partition being stolen by another event consumer asserting exclusive access, the failure must be surfaced to callers so that the correct remediation can be taken.

The focus of these changes is to ensure that errors causing link termination are surfaced via logging and that a stolen partition is treated as a special case and is always surfaced to callers.  Because this behavior was a direct impact to the load balancing aspects of the event processor, testing exposed some additional load balancing gaps that were also addressed.  Lost partitions will now be detected more reliably and the processor will proactively relinquish ownership and allow load balancing to make recovery decisions.

# References and Resources 

- [Event Hubs: Fix Unobserved Stolen Partitions (#21297)](https://github.com/Azure/azure-sdk-for-net/issues/21297)

### CODE REVIEW

File: sdk/eventhub/Azure.Messaging.EventHubs/api/Azure.Messaging.EventHubs.netstandard2.0.cs [hunk 1]
```diff
@@ -371,6 +371,7 @@ public abstract partial class EventProcessor<TPartition> where TPartition : Azur
         protected virtual System.Threading.Tasks.Task OnPartitionProcessingStoppedAsync(TPartition partition, Azure.Messaging.EventHubs.Processor.ProcessingStoppedReason reason, System.Threading.CancellationToken cancellationToken) { throw null; }
         protected abstract System.Threading.Tasks.Task OnProcessingErrorAsync(System.Exception exception, TPartition partition, string operationDescription, System.Threading.CancellationToken cancellationToken);
         protected abstract System.Threading.Tasks.Task OnProcessingEventBatchAsync(System.Collections.Generic.IEnumerable<Azure.Messaging.EventHubs.EventData> events, TPartition partition, System.Threading.CancellationToken cancellationToken);
+        protected virtual System.Threading.Tasks.Task<string[]> QueryPartitionIds(Azure.Messaging.EventHubs.EventHubConnection connection, System.Threading.CancellationToken cancellationToken) { throw null; }
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/api/Azure.Messaging.EventHubs.netstandard2.0.cs:374]: This is orthogonal, but had been discussed with the architects at an earlier point in time.  This allows custom processors to more easily assign a specific set of partitions for the processor to own.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpConnectionScope.cs [hunk 2]
```diff
@@ -240,17 +240,31 @@ protected AmqpConnectionScope()
...
 
-            var link = await CreateManagementLinkAsync(connection, timeout.CalculateRemaining(stopWatch.GetElapsedTime()), cancellationToken).ConfigureAwait(false);
-            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+                var stopWatch = ValueStopwatch.StartNew();
+                var connection = await ActiveConnection.GetOrCreateAsync(timeout).ConfigureAwait(false);
+                cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
 
-            await OpenAmqpObjectAsync(link, timeout.CalculateRemaining(stopWatch.GetElapsedTime())).ConfigureAwait(false);
-            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+                var link = await CreateManagementLinkAsync(connection, timeout.CalculateRemaining(stopWatch.GetElapsedTime()), cancellationToken).ConfigureAwait(false);
```
Comment (christothes) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpConnectionScope.cs:251]: Just curious - How is this timespan used? It seems inherently racy to be calculating this as we go rather than relying on a cancellation token or reference to a known expiration DateTime.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpConsumer.cs [hunk 3]
```diff
@@ -444,6 +508,61 @@ public override async Task CloseAsync(CancellationToken cancellationToken)
...
+            {
+                return;
+            }
+
+            // If the consumer is being closed, the sentinel variable will already be set, and
+            // the terminal exception need not be considered.
+
+            if (!_closed)
+            {
+                var linkException = GetTerminalException(link);
```
Comment (christothes) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpConsumer.cs:531]: nit: this could be simplified to:
```suggestion
                var linkException = link?.TerminalException;
```

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Diagnostics/EventHubsEventSource.cs [hunk 4]
```diff
@@ -1093,6 +1093,308 @@ public virtual void ReadAllEventsComplete(string eventHubName)
...
+        /// </summary>
+        ///
+        /// <param name="objectTypeName">The name of the AMQP object type being managed; normally this will be a type of AMQP link.</param>
+        /// <param name="identifier">The identifier of the client associated with the AMQP object.</param>
+        /// <param name="eventHubName">The name of the Event Hub that the AMQP object is associated with.</param>
+        /// <param name="consumerGroup">The name of the consumer group that the AMQP object is associated with.</param>
+        /// <param name="partitionId">The identifier of the partition that the AMQP object is associated with.</param>
+        /// <param name="errorMessage">The message for any terminal exception that may have occurred.</param>
+        ///
+        [Event(55, Level = EventLevel.Verbose, Message = "An AMQP object of type '{0}' was closed by the fault tolerant manager for client '{1}', Event Hub: {2}, Consumer Group: {3}, and partition: {4}.  Terminal error message: '{5}'")]
```
Comment (christothes) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Diagnostics/EventHubsEventSource.cs:1107]:
```suggestion
        [Event(55, Level = EventLevel.Verbose, Message = "An AMQP object of type '{0}' was closed by the fault tolerant manager for client: '{1}', Event Hub: {2}, Consumer Group: {3}, and partition: {4}.  Terminal error message: '{5}'")]
```

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Diagnostics/EventHubsEventSource.cs [hunk 5]
```diff
@@ -1093,6 +1093,308 @@ public virtual void ReadAllEventsComplete(string eventHubName)
...
+        }
+
+        /// <summary>
+        ///   Indicates that creating an AMQP connection has started.
+        /// </summary>
+        ///
+        /// <param name="endpoint">The service endpoint that the connection is being opened for.</param>
+        /// <param name="transportType">The type of transport being used for the connection</param>
+        ///
+        [Event(56, Level = EventLevel.Verbose, Message = "Beginning creation of an AMQP connection for '{0}' using the transport '{1}`.")]
```
Comment (christothes) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Diagnostics/EventHubsEventSource.cs:1128]: nit for consistency:
```suggestion
        [Event(56, Level = EventLevel.Verbose, Message = "Beginning creation of an AMQP connection for endpoint: '{0}' using the transport: '{1}`.")]
```

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor.cs [hunk 6]
```diff
@@ -43,6 +43,9 @@ public abstract class EventProcessor<TPartition> where TPartition : EventProcess
         /// <summary>The maximum number of failed consumers to allow when processing a partition; failed consumers are those which have been unable to receive and process events.</summary>
         private const int MaximumFailedConsumerCount = 1;
 
+        /// <summary>Indicates whether or not the consumer should consider itself invalid when a partition is stolen by another consumer, as determined by the Event Hubs service.</summary>
+        private const bool InvalidateConsumerWhenPartitionIsStolen = true;
```
Comment (christothes) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor.cs:47]: I see this defaulted to false elsewhere. Is that expected?

**Review by christothes — APPROVED** [review]

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpConnectionScope.cs (hunk 2 above)
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpConnectionScope.cs:251]: The AMQP library requires an independent timeout value (as `TimeSpan`)  for each and every operation.  Its the responsibility of the caller (us, in this case) to take the `TryTimeout` and reduce that as an overarching operation timeout applied across any and all AMQP interactions.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor.cs (hunk 6 above)
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor.cs:47]: Yup;  we want the other consumer and receiver clients to fail if they ask to read events so they see the background exception.   The caller can make the decision to invoke again which will reassert control when competing consumers have the same owner level.   _(in the future, we may expose this as an option for the other types)_

In this case, we want the partition processor to fault and not have a chance to recover with retries so that it immediately respects the ownership change for the partition.  In the worst case, when load balancing runs again, if storage says this processor still owns the partition, a new partition processing task will get spun up and reassert ownership.

In truth, it's a bit of overkill, as we don't normally retry this type of exception, but we have the mechanism in place and using it makes sense.
